### PR TITLE
JS: Extract XML with codeql when run with codeql

### PR DESCRIPTION
**Pull request by asgerf**
XML extraction currently works by AutoBuild calling back to `odasa` to invoke the XML extractor. This now calls back to `codeql` when `CODEQL_DIST` is set.

It's questionable whether it _should_ be done like this, but given that our tests are blocked on this, I opted for the minimal change needed so we can get our tests running again.

### CODE REVIEW

File: javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java [hunk 1]
```diff
@@ -1239,11 +1240,29 @@ private void logEndProcess(long timedLogMessageStart, String message) {
...
+    if (EnvironmentVariables.getCodeQLDist() == null) {
+      // Use the legacy odasa XML extractor
+      cmd.add("odasa");
+      cmd.add("index");
+      cmd.add("--xml");
+      cmd.add("--extensions");
+      cmd.addAll(xmlExtensions);
+    } else {
+      String command = Env.getOS() == OS.WINDOWS ? "codeql.cmd" : "codeql";
+      cmd.add(EnvironmentVariables.getCodeQLDist() + "/" + command);
```
Comment (esbena) [inline javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java:1252]: Shouldnt `"/"` be something like `File.separator`?
 (or maybe just use something like `Path.resolve` instead, as done elsewhere in this file)

**Review by esbena — CHANGES_REQUESTED** [review]

File: javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java [hunk 2]
```diff
@@ -1239,11 +1240,29 @@ private void logEndProcess(long timedLogMessageStart, String message) {
...
-    cmd.addAll(xmlExtensions);
+    if (EnvironmentVariables.getCodeQLDist() == null) {
+      // Use the legacy odasa XML extractor
+      cmd.add("odasa");
+      cmd.add("index");
+      cmd.add("--xml");
+      cmd.add("--extensions");
+      cmd.addAll(xmlExtensions);
+    } else {
+      String command = Env.getOS() == OS.WINDOWS ? "codeql.cmd" : "codeql";
```
Comment (aibaars) [inline javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java:1251]: I think `codeql.cmd` is deprecated in favour of `codeql.exe`.

File: javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java (hunk 1 above)
Comment (aibaars) [reply javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java]: I'd be in favour of using `Paths.get(EnvironmentVariables.getCodeQLDist(), command)` or indeed `Paths.resolve()`

**Review by esbena — DISMISSED** [review]
LGTM.

**Review by aibaars — COMMENTED** [review]
This looks good as a quick solution. I think it would be best to move XML indexing as a `pre-finalize` script. That way the CodeQL CLI can check whether there were any source files extracted by the main autobuild and fail extraction before indexing additional files.

To implement that properly we should add an option (environment variable) to the XML extractor so you can supply the desired extensions. This option could be set by the `lgtm.yml` interpreter.

How to supply additional options to extractors is still a draft design by @p0 , so you might want to wait a little before implementing this.

**Review by aibaars — APPROVED** [review]

**Review by esbena — APPROVED** [review]

**Comment by asgerf** [comment]
Thanks. I'm still waiting for the internal JS-AutoBuild tests to pass (not part of the PR checks afaik).

**Pull request merged**